docs/apache-airflow/administration-and-deployment/logging-monitoring/advanced-logging-configuration.rst

```diff
@@ -0,0 +1,79 @@
+
+ .. Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+
+
+Advanced logging configuration
+------------------------------
+
+Not all configuration options are available from the ``airflow.cfg`` file. The config file describes
+how to configure logging for tasks, because the logs generated by tasks are not only logged in separate
+files by default but has to be also accessible via the webserver.
+
+By default standard airflow component logs are written to the ``$AIRFLOW_HOME/logs`` directory, but you
+can also customize it and configure it as you want by overriding Python logger configuration that can
+be configured by providing custom logging configuration object. Some configuration options require
+that the logging config class be overwritten. You can do it by copying the default
+configuration of Airflow and modifying it to suit your needs. The default configuration can be seen in the
+`airflow_local_settings.py template <https://github.com/apache/airflow/blob/|airflow_version|/airflow/config_templates/airflow_local_settings.py>`_
+and you can see the loggers and handlers used there. Except the custom loggers and handlers configurable there
+via the ``airflow.cfg``, the logging methods in Airflow follow the usual Python logging convention,
+that Python objects log to loggers that follow naming convention of ``<package>.<module_name>``.
+
+You can read more about standard python logging classes (Loggers, Handlers, Formatters) in the
+`Python logging documentation <https://docs.python.org/3/library/logging.html>`_.
+
+Configuring your logging classes can be done via the ``logging_config_class`` option in ``airflow.cfg`` file.
+This configuration should specify the import path to a configuration compatible with
+:func:`logging.config.dictConfig`. If your file is a standard import location, then you should set a
+:envvar:`PYTHONPATH` environment variable.
+
+Follow the steps below to enable custom logging config class:
+
+#. Start by setting environment variable to known directory e.g. ``~/airflow/``
+
+    .. code-block:: bash
+
+        export PYTHONPATH=~/airflow/
+
+#. Create a directory to store the config file e.g. ``~/airflow/config``
+#. Create file called ``~/airflow/config/log_config.py`` with following the contents:
+
+    .. code-block:: python
+
+      from copy import deepcopy
+      from airflow.config_templates.airflow_local_settings import DEFAULT_LOGGING_CONFIG
+
+      LOGGING_CONFIG = deepcopy(DEFAULT_LOGGING_CONFIG)
+
+#.  At the end of the file, add code to modify the default dictionary configuration.
+#. Update ``$AIRFLOW_HOME/airflow.cfg`` to contain:
+
+    .. code-block:: ini
+
+        [logging]
+        remote_logging = True
```

Comment:
Just curious if we need `remote_logging` to be set to `True` for modifying the `logging_config_class`? If not perhaps, we can remove this keypair from here?

Reply:
Good point. Added a bit more elaborated description and explained both cases - with and without remote logging.

The thing here is that the deepcopied configuration will contain / not contain remote logging loggers and handlers depending on the remote_logging setting - the dictionary of logger configuration is calculated dynamically when airflow settings are parsed, and I can easily imagine users that could choose advanced configuration based on "plain" airflow logging or the one that is calculated when remote logging is enabled.

And yeah, it is confusing, because on one hand, you can modify the dictionary for logging based on declarative "config" but then you can modify the configuration dynamically right after it is calculated. It's been always like that and it is complex on one hand, but on the other hand, it combines the simplicity and declarative way of remote logging configuration for tasks, with an absolute flexibility of configuring logging in whatever way suits the final user. It also allows us to keep the configuration changes that should be applied to everyone (for example in 2.4 I think @dstandish changed the way how handlers are propagated, so keeping the setting that users will modify and deepcopy calculated configuration rather than just copy the whole of it is the only "reasonable" (even if complex) to also allow custom configuration working across version upgrades of airlfow.